ykrt/src/compile/jitc_yk/codegen/x86_64.rs

```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    fn reg_into_new_local(&mut self, local: InstrIdx, reg: Rq) {
+        let l = self.allocate_local(local);
+        self.store_local(&l, reg, local.instr(self.jit_mod).def_abi_size())
+    }
+
+    pub(super) fn codegen_loadarg_instr(
+        &mut self,
+        inst_idx: jit_ir::InstrIdx,
+        _inst: &jit_ir::LoadArgInstruction,
+    ) {
+        // FIXME: LoadArg instructions are not yet specified. This hack is just to satisfy data
+        // flow dependencies during code generation.
+        dynasm!(self.asm ; mov Rq(WR0.code()), 0);
+        self.reg_into_new_local(inst_idx, WR0);
+    }
+
+    pub(super) fn codegen_load_instr(
+        &mut self,
+        inst_idx: jit_ir::InstrIdx,
+        inst: &jit_ir::LoadInstruction,
+    ) {
+        self.operand_into_reg(WR0, &inst.operand());
+        let size = inst_idx.instr(self.jit_mod).def_abi_size();
+        debug_assert!(size <= REG64_SIZE);
+        match size {
+            8 => dynasm!(self.asm ; mov Rq(WR0.code()), [Rq(WR0.code())]),
+            4 => dynasm!(self.asm ; mov Rd(WR0.code()), [Rq(WR0.code())]),
+            2 => dynasm!(self.asm ; movzx Rd(WR0.code()), WORD [Rq(WR0.code())]),
+            1 => dynasm!(self.asm ; movzx Rq(WR0.code()), BYTE [Rq(WR0.code())]),
+            _ => todo!("{}", size),
+        };
+        self.reg_into_new_local(inst_idx, WR0);
+    }
+
+    pub fn const_u64_into_reg(&mut self, reg: Rq, cv: u64) {
+        dynasm!(self.asm
+            ; mov Rq(reg.code()), QWORD cv as i64 // `as` intentional.
+        )
+    }
+
+    /// Load an operand into a register.
+    pub fn operand_into_reg(&mut self, reg: Rq, op: &Operand) {
+        //if v.abi_size_in_bytes(self.target) > REG64_SIZE {
+        //    todo!();
+        //}
+
+        match op {
+            Operand::Local(li) => self.load_local(reg, *li),
+            _ => todo!("{}", op),
+        }
+    }
+}
+
+#[cfg(any(debug_assertions, test))]
+impl AsmPrinter<'_> {
+    pub(super) fn target_print(
+        &self,
+        out: &mut Vec<String>,
+        code: &[u8],
+        start_vaddr: u64,
+        len: usize,
+    ) {
+        let mut dis = iced_x86::Decoder::with_ip(64, code, start_vaddr, 0);
+        let mut remain = len;
+        while remain != 0 {
+            let off = len - remain;
+            if let Some(comments) = self.comments {
+                if let Some(lines) = comments.get(&AssemblyOffset(off)) {
+                    for line in lines {
+                        out.push(format!("; {line}"));
+                    }
+                }
+            }
+            let inst = dis.decode();
+            out.push(format!("{:08x} {:08x}: {}", inst.ip(), off, inst));
+            remain -= inst.len();
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::CodeGen;
+    use crate::compile::jitc_yk::{aot_ir, codegen::tests::match_asm, jit_ir};
+
+    #[test]
+    fn simple_codegen() {
+        let mut aot_mod = aot_ir::Module::default();
```

Comment:
@ltratt @ptersilie  take a look at this test. Here we have to build an AOT module so that the JIT module can refer to parts of it. It's a bit awkward.

We could make the JIT module independent of the AOT module at the cost of copying. Is it worth the performance hit for testing?

Reply:
> Is it worth the performance hit for testing?

If it's the difference between "testing is intolerable" and "testing is tolerable" then I think "yes".

Reply:
@ptersilie what do you think?

Reply:
I dunno. I seems a bit backwards to me to affect your performance for the sake of making testing easier. But if it makes us test less because it's such a hassle then that's also bad. Maybe we could have a general AOT module that we then use for all the tests. Or we make a helper that makes it easier to construct the AOT module. Maybe some syntax where you make a trace and it then generates the corresponding AOT module automatically.

I dunno, since @ltratt was so adamant we never ever copy from the AOT if we can help it, I think he will have to make that decision.

Reply:
I didn't think about the impact of testing. Given the choice between "make testing plausible" and "squeeze every last drop of performance" I would go for "make testing plausible".

Reply:
Some things, e.g. constant strings, constant arrays, etc, will be very expensive to copy though.

Reply:
They won't be *very* expensive :) And at the moment testability is going to save *us* more time.

Reply:
OK, well that's definitely a separate PR, as there are many tentacles.

It will also cause more of the "search for this thing, or insert it" idiom, which @ltratt has expressed distaste for.

For example, types will no longer be shared between AOT and JIT modules, and there will need to be a mapping between the now separate elements in each module kind. When you have a handle on an AOT type and want to get the corresponding JIT type, can we avoid that idiom somehow?

Reply:
> there will need to be a mapping between the now separate elements in each module kind

I don't get this bit.

Reply:
So that's a no on generating the AOT/JIT module automatically from some description for testing?

Reply:
> > there will need to be a mapping between the now separate elements in each module kind
> 
> I don't get this bit.

I think he means that in order to not copy the same type multiple types you need to know if you've already copied it and then return the copy instead. To do that you probably need some kind of mapping, e.g. `Types[AOTType] = JITType`

Reply:
> So that's a no on generating the AOT/JIT module automatically from some description for testing?

It sounds like it could be fiddly to be honest.

Reply:
We could use a hashmap, yes, but I thought you wouldn't like it. If that's OK then it simplifies some other paces where that idiom is in use too (separate PR).

Reply:
The other possibility is to have two modes: a testing mode where we copy everything; and a non-testing mode where we reference things. Ultimately "make it testable" is our goal. We won't get everything right, but if that's our goal, at least we'll know what to prioritise.

Reply:
OK, so I think we should roll will "make it testable". I've made a note for two future PRs based on this discussion. Thanks.